src/Microsoft.ML.Data/Model/ModelOperationsCatalog.cs

```diff
@@ -28,64 +28,69 @@ internal ModelOperationsCatalog(IHostEnvironment env)
+            _env.CheckValueOrNull(model);
             _env.CheckValue(stream, nameof(stream));
 
+            // For the sake of consistency of this API specifically, when called upon we save any transformer
+            // in a single element transformer chain.
+            var chainedModel = model == null ? null : new TransformerChain<ITransformer>(model);
+            var compositeLoader = new CompositeDataLoader<TSource, ITransformer>(loader, chainedModel);
+
             using (var rep = RepositoryWriter.CreateNew(stream))
             {
-                ModelSaveContext.SaveModel(rep, model, null);
+                ModelSaveContext.SaveModel(rep, compositeLoader, null);
                 rep.Commit();
             }
         }
 
-        /// <summary>
-        /// Save the model to the file.
-        /// </summary>
-        /// <param name="model">The trained model to be saved.</param>
-        /// <param name="filePath">Path where model should be saved.</param>
-        public void Save<TSource>(IDataLoader<TSource> model, string filePath)
-        {
-            using (var stream = File.Create(filePath))
-                Save(model, stream);
-        }
-
-        /// <summary>
-        /// Save a transformer model and the loader used to create its input data to the stream.
-        /// </summary>
-        /// <param name="loader">The loader that was used to create data to train the model</param>
-        /// <param name="model">The trained model to be saved</param>
-        /// <param name="stream">A writeable, seekable stream to save to.</param>
-        public void Save<TSource>(IDataLoader<TSource> loader, ITransformer model, Stream stream) =>
-            Save(new CompositeDataLoader<TSource, ITransformer>(loader, new TransformerChain<ITransformer>(model)), stream);
-
         /// <summary>
         /// Save a transformer model and the loader used to create its input data to the file.
         /// </summary>
-        /// <param name="loader">The loader that was used to create data to train the model</param>
-        /// <param name="model">The trained model to be saved</param>
+        /// <param name="model">The trained model to be saved. Note that this can be <see langword="null"/>, as a shorthand
+        /// for an empty data set. Upon loading with <see cref="LoadWithDataLoader(Stream, out IDataLoader{IMultiStreamSource})"/>
+        /// the returned valued will be an empty <see cref="TransformerChain{TLastTransformer}"/>.</param>
+        /// <param name="loader">The loader that was used to create data to train the model.</param>
         /// <param name="filePath">Path where model should be saved.</param>
-        public void Save<TSource>(IDataLoader<TSource> loader, ITransformer model, string filePath)
+        public void Save<TSource>(ITransformer model, IDataLoader<TSource> loader, string filePath)
         {
+            _env.CheckValueOrNull(model);
+            _env.CheckValue(loader, nameof(loader));
```

Comment:
One thing that I find a little bit odd about this behavior is when a user calls:

```
ITransformer model = ...;
mlContext.Model.Save(model, null, filePath);
```

The `null` argument in the 2nd argument is going to be ambiguous between the two overloads, and they will need to explicitly pick one.

If they choose the `DataViewSchema` overload, the above code will succeed.
If they choose the `IDataLoader` overload, it will throw an ArgumentNullException.

This probably isn't a huge issue. Just something I thought I'd bring up for discussion.

I almost wonder if we should have a `mlContext.ModelSave(model, filePath)` overload that doesn't require the loader or schema.

Reply:
> This probably isn't a huge issue. Just something I thought I'd bring up for discussion.

It is interesting @eerhardt, but let's remember, as per the discussion of #2663, that this scenario is extremely rare. Most of the time people want to save a complete model, not just bits and pieces of it. (I *did* notice that though, and we do have test code that exercises it as you will see, and due to the way the overload resolution works, it does in fact do the right thing.)

> I almost wonder if we should have a `mlContext.ModelSave(model, filePath)` overload that doesn't require the loader or schema.

That would be a repetition of the exact mistake that led to #2663, and the necessary changes in this API. I *want* the suggestion that the input schema (whether by the schema or the loader that generates the schema) to be right in the user's face. We know what happens when that suggestion is not there, because the situation where you don't want to do that is actually quite rare.

[I had a comment](https://github.com/dotnet/machinelearning/pull/2858#discussion_r262992588) in #2858 where I brought up why this was a mistake... to summarize and rephrase, the danger is, if you have a simpler looking API, people will tend to *use* that simpler looking API. (Because, heck, it looks simpler.) That's *very* dangerous if that simpler looking API does something that is wrong in 99.9% of situations. Even more dangerous if it's not *obvious* it is doing the wrong thing -- and this was *so* non-obvious that it was doing the wrong thing that @Ivanidzo4ka filed an issue and even went so far as to label it as an actual bug, when he's probably one of the people that knows our architecture the best.

No, this suggestion, "here's where you save the starting point of your model," is a really, really important thing, and I would consider acting on this specific suggestion a mistake.

Reply:
> because the situation where you don't want to do that is actually quite rare.

> That's very dangerous if that simpler looking API does something that is wrong in 99.9% of situations.

The part of this whole situation that has me confused is:

How was anything working previously? People are saving models today without the loader/schema, and they are successfully using the models to make predictions in their applications.

I get that there are scenarios where you will need the loader/schema information. But is it _really_ 99.9% of the time where you _need_ these things? How are people successfully using the Save/Load APIs today without it? Are they really into the 0.1% use case?

Reply:
I get that we want to lead people down a "pit of success" - which is ensuring that they at least know that they should be saving the loader/schema. So I'll resolve this conversation. I get the concern. Thanks for the discussion.

Reply:
So, I am on my phone so I can't find them now. The fact that we were failing to save the data loader has been a problem since the v0.1 API.

The fact that it appears to work sometimes is what makes it so pernicious. You don't need the input schema until you do, then if you don't have it you can't do simple things like figure out how many features a model expects, what the names were, etc. Then that information is not there unless you reconstruct the input... Which all of our samples happen to do.

Being able to get the output schema of a model without knowing what the input was is really important...